Add ability to toggle mods during replay

cdwcgt opened the pull request:
close https://github.com/ppy/osu/issues/8357

create a new interface named `ICanBeToggledDuringReplay`
Some annotations are missing, probably

add a test for mod can only toggle in replay

~~Maybe should handle `ReplayLoaded` in `HudOverlay`~~ Handle at [ClickableModIcon.cs](https://github.com/ppy/osu/pull/19925/files#diff-99f16b384191bb1d0acd55a451eb60169f5f5a73adfd12ad0ec9d283e31cc650R35)

mod list that I think can be implemented
- ~~[ ] HR~~ (Unable to implement, need to reload `Player`, also replay may broken)
- [x] Flashlight
- [x] Binds
- [x] RateAdjust (DT HT NC DC)
- [x] TimeRamp (Wind up/down)
- [x] ModAdaptiveSpeed
- [x] Hidden (``ModWithVisibilityAdjustment``) 
- [x] NoScope
- [x] Muted
- [x] Floating Fruits
- [x] BarrelRoll
- [x] OsuModApproachDifferent
- [x] OsuModFreezeFrame
- ~~[ ] ManiaModConstantSpeed~~ (Unable to implement, need to reload `DrawableScrollingRuleset`)
- [x] OsuModBubbles
- [x] ModSynesthesia
- [x] OsuModDepth (maybe not suitable to disable, replay may broken)

cdwcgt committed:
Add ability to allow mods to disable during replay

cdwcgt committed:
simple implement for OsuModFlashlight

cdwcgt committed:
Prevent tests from hitting `ModIcon`

a poor workaround
but I don't know how to solve it better

cdwcgt committed:
rename `IApplicableToDisable` to `ICanBeToggledDuringReplay`

cdwcgt committed:
fix default value

cdwcgt committed:
Add test for mod can only toggle in replay

cdwcgt committed:
implement `ICanBeToggledDuringReplay`  for `OsuModBlinds`

cdwcgt committed:
implement `ICanBeToggledDuringReplay`  for `ModRateAdjust`

cdwcgt committed:
implement `ICanBeToggledDuringReplay`  for `ModTimeRamp`

bdach commented on osu.Game/Rulesets/Mods/ICanBeToggledDuringReplay.cs:
```diff
@@ -0,0 +1,22 @@
+{
+    public interface ICanBeToggledDuringReplay : IApplicableMod
+    {
+        Bindable<bool> ReplayLoaded
+        {
+            get;
+        }
+
+        bool IsDisable
+        {
+            get;
+        }
+
+        void OnToggle();
```
* No xmldoc, no description of what the interface is for or what the individual members do. This is not acceptable for a ruleset API.
* Why is `ReplayLoaded` in this interface? Ideally it would not be.
* What is `IsDisable` even supposed to be?

bdach commented on osu.Game/Rulesets/UI/ModIcon.cs:
```diff
@@ -96,6 +96,16 @@ public ModIcon(IMod mod, bool showTooltip = true)
                 },
             };
+
+            Action = () =>
+            {
+                if (mod is ICanBeToggledDuringReplay dmod)
+                {
+                    dmod.OnToggle();
+
+                    Colour = dmod.IsDisable ? OsuColour.Gray(0.7f) : Colour = Color4.White;
+                }
+            };
```
This cannot be done in `ModIcon`. The class is not only used in the replay player. Behaviours specific to a single screen cannot and should not be added to globally-used classes, it's just bad layering.

bdach commented on osu.Game.Tests/Visual/Multiplayer/TestSceneDrawableRoomPlaylist.cs:
```diff
@@ -128,7 +128,8 @@ public void TestEditableSelectable()
             });
 
-            moveToItem(0);
+            // ModIcon will handle clicks before EditableSelectable
+            moveToItem(0, new Vector2(50, 0));
```
This is not an acceptable way to fix any tests and precisely demonstrates why setting `ModIcon.Action` was as bad an idea as I'm saying it was.

bdach reviewed (changes requested):
I have no idea what anything is here.

cdwcgt committed:
Create `ClickableModDisplay` and `ClickableModIcon` to instead of `ModIcon`

cdwcgt committed:
add description to interface

cdwcgt committed:
Revert "Prevent tests from hitting `ModIcon`"

This reverts commit cbfcaab74a46a9d2502eb31a22cca3e0d3dcba28.

cdwcgt committed:
update test for new `ClickableModIcon` class

cdwcgt committed:
make `replayLoaded` handle by `ClickableModIcon`

cdwcgt committed:
remove blank line

my bad

peppy commented on osu.Game/Rulesets/Mods/ICanBeToggledDuringReplay.cs:
```diff
@@ -0,0 +1,24 @@
+    /// </summary>
+    public interface ICanBeToggledDuringReplay : IApplicableMod
+    {
+        /// <summary>
+        /// A property to set whether the mod has been disabled.
+        /// </summary>
+        bool IsDisable
+        {
+            get;
+        }
```
Without digging too deep into this, this doesn't make sense. A get property to set?

peppy commented on osu.Game/Rulesets/Mods/ICanBeToggledDuringReplay.cs:
```diff
@@ -0,0 +1,24 @@
+        /// <summary>
+        /// A property to set whether the mod has been disabled.
+        /// </summary>
+        bool IsDisable
```
`Disabled` or `IsDisabled`

peppy commented on osu.Game/Rulesets/Mods/ICanBeToggledDuringReplay.cs:
```diff
@@ -0,0 +1,24 @@
+        /// <summary>
+        /// Method called when mod toggle.
+        /// </summary>
+        void OnToggle();
```
You may just want one single `BindableBool` in this interface, which can handle both value and event flow.

cdwcgt replied:
Sure, but I'm a little confused where the event is registered
There are some mods that don't have a constructor

Define an event registration method in an interface?

bdach replied:
What "event"? You mean the `BindValueChanged` callback of the proposed bindable in question?

Mods that don't *currently* have a constructor can receive one. I'm not sure what the problem is here.

cdwcgt replied:
I have decided to use a method to register events and this method will be called once on load

peppy commented on osu.Game/Rulesets/UI/ClickableModIcon.cs:
```diff
@@ -0,0 +1,151 @@
+﻿// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+#nullable disable
```
Please no `nullable disable` on new classes

cdwcgt replied:
These two class is copied from `ModIcon` and `ModDisplay`

I think we can use my processing method in these two classes

But not within the scope of pr, I didn't change them

cdwcgt committed:
Change the toggle logic

bring in `DisableToggleEvent` to register event handle for `BindableBool IsDisabled`

cdwcgt committed:
remove #nullable disable

cdwcgt committed:
Update Test for change of interface

cdwcgt committed:
inheritance reuse

cdwcgt committed:
Create a local bindable in `DrawableOsuBlinds`

cdwcgt committed:
remove `DisableToggleEvent()`

cdwcgt committed:
Merge remote-tracking branch 'upstream/master' into mod-disable

cdwcgt committed:
implement `ICanBeToggledDuringReplay` for `ModeFlashlight`

Can provide a base implementation for all mods that inherit `ModeFlashlight`

cdwcgt committed:
implement `ICanBeToggledDuringReplay` for `ManiaModPlayfieldCover`

cdwcgt commented:
I'm having trouble switching hidden mods in other ruleset (e.g stable, taiko, catch)
Their hidden is implemented by FadeOut() or Hide()
and I have no idea to revert this.

Hope someone can help me with this

cdwcgt committed:
Merge remote-tracking branch 'upstream/master' into mod-disable

cdwcgt committed:
implement `ICanBeToggledDuringReplay` for `ModAdaptiveSpeed`

cdwcgt committed:
implement `ICanBeToggledDuringReplay` for `CatchModFloatingFruits`

cdwcgt committed:
implement `ICanBeToggledDuringReplay` for `ModNoScope`

cdwcgt committed:
implement `ICanBeToggledDuringReplay` for `ModMuted`

cdwcgt committed:
optimization logic

cdwcgt commented:
> I'm having trouble switching hidden mods in other ruleset (e.g stable, taiko, catch) Their hidden is implemented by FadeOut() or Hide() and I have no idea to revert this.
> 
> Hope someone can help me with this

I‘m fool

cdwcgt committed:
Code Quality fix

cdwcgt committed:
implement `ICanBeToggledDuringReplay` for `ModWithVisibilityAdjustment`

cdwcgt committed:
implement `ICanBeToggledDuringReplay` for `ModBarrelRoll`

cdwcgt committed:
implement `ICanBeToggledDuringReplay` for `OsuModApproachDifferent`

cdwcgt committed:
the better way to implement disable `OsuModApproachDifferent`

frenzibyte commented on osu.Game.Rulesets.Osu/Objects/Drawables/DrawableHitCircle.cs:
```diff
@@ -172,6 +172,7 @@ protected override void UpdateInitialTransforms()
             CirclePiece.FadeInFromZero(HitObject.TimeFadeIn);
 
             ApproachCircle.FadeIn(Math.Min(HitObject.TimeFadeIn * 2, HitObject.TimePreempt));
+            ApproachCircle.ScaleTo(4);
             ApproachCircle.ScaleTo(1f, HitObject.TimePreempt);
```
```suggestion
            ApproachCircle.ScaleTo(4).ScaleTo(1f, HitObject.TimePreempt);
```

cdwcgt commented:
Looks like basically all mods that can do it do it I think

Most of the other mods directly modify the beatmap instead of the visual effect, so I will not consider it for the time being.

cdwcgt committed:
Optimize the ScaleTo method in `DrawableHitCircle.UpdateInitialTransforms()`

Co-authored-by: Salman Ahmed <frenzibyte@gmail.com>

bdach commented:
I'm just going to go on record and state that I'm not sure that this feature should exist, at least not given how it's currently implemented. The `IsDisabled` checks strewn all over existing mods are really quite ugly, and all future implementors of all future mods will have to remember to explicitly opt into this behaviour if they are to handle toggling during replay.

I don't really have any concrete thoughts on what the ideal implementation would be, aside from one constraint, that there ideally should be no conditionals in mod logic. A marker interface for mods that are toggleable would be fine (because as stated, some mods don't make sense to toggle during a replay), but ideally no mod logic should be explicitly handling toggle.

As for how to achieve this - your guess is as good as mine, but it'd probably require a different angle of attack (rather than putting the burden on toggling mod effects onto the mods, it would have to originate somewhere in `ReplayPlayer`).

cdwcgt commented:
The specific behavior of some mods requires specific methods to roll them back (such as IApplyTo DrawableRuleset)
Define a method through an interface and call it in ReplayPlayer?

smoogipoo commented:
I'm not sure how relevant it is here, but we've discussed having "abilities" in multiplayer that you can enable for a short amount of time on opponents. Not sure if mods are the best way to implement these abilities, but I could totally see enabling barrel roll mod for a few seconds as an example, which requires implementation outside of `ReplayPlayer`.

peppy commented:
Originally we had the idea of `ApplyMod` and `RevertMod` implemented by each mod, but it turned out to be a bit complicated implementation wise. Definitely needs further experimentation. I'm not necessarily against the direction this PR is taking, but due to the scope of the PR I'm not sure it will ever get merged.

I'd want to see it implemented for just one mod to begin with, so we can experiment with changes to implementation without having changes across many files.

cdwcgt commented:
So open a pr only to implement ICanBeToggleDuringReplay interface and implement one of mods (I wish it can be flashlight)
then implement other mods in other pr?

peppy commented:
That might be best, yes. You can make this PR dependent on that and leave it in its current state if you wish.

Keep in mind this is going to be a pretty low priority thing as the time required to think about the correct forward direction is non-negligible.

cdwcgt commented:
ok, I decided to wait for your results

cdwcgt committed:
Merge branch 'master' into mod-disable

cdwcgt committed:
Merge remote-tracking branch 'upstream/master' into mod-disable

cdwcgt committed:
implement `ICanBeToggledDuringReplay` for `OsuModFreezeFrame`

cdwcgt committed:
Merge branch 'master' into mod-disable

cdwcgt committed:
make partial

cdwcgt committed:
update color when the `mod.IsDisabled` is changed externally

https://github.com/ppy/osu/pull/19925#issuecomment-1238912446
Assuming changing the mod state in multiplayer, ModIcon will also update its color

cdwcgt committed:
make partial

cdwcgt committed:
Merge branch 'master' into mod-disable

cdwcgt committed:
implement `ICanBeToggledDuringReplay` for `OsuModBubbles`

cdwcgt committed:
Merge branch 'master' into mod-disable

cdwcgt committed:
fix nightcore freqAdjust not revert when disable

Keep `NightcoreBeatContainer` as the same as stable do.

cdwcgt committed:
fix disable freeze frame using the wrong Time Preempt

this problem cause disable freeze frame ignore beatmap AR

cdwcgt committed:
implement `ICanBeToggledDuringReplay` for `OsuModSynesthesia`

cdwcgt committed:
fix Daycore Nightcore have wrong Tempo apply for track when disabled

cdwcgt committed:
Merge branch 'master' into mod-disable

cdwcgt committed:
remove launchSettings

cdwcgt committed:
add comment for `ApproachCircle` transform change

not sure if it will affect performance

cdwcgt committed:
fix rate adjust

cdwcgt committed:
implement `ICanBeToggledDuringReplay` for `OsuModDepth`

It may should cause the replay to play in an unexpected way, example hit the circle before it arrive at location

Maybe this mod is not suitable for turning it off.

cdwcgt committed:
format

cdwcgt committed:
fix floating fruits

cdwcgt committed:
fix TimeRamp

cdwcgt committed:
add comment for `DisableSpeedChange` in `RateAdjustModHelper`

cdwcgt committed:
Merge branch 'master' into mod-disable

cdwcgt committed:
fix disable function for cover mod
